src/tex/crunch_texture.cpp

```diff
@@ -0,0 +1,192 @@
+                           &mipmapped_texture::get_total_pixels);
+    if (!first)
+        return false;
+
+    image_u8 lhs_temp, rhs_temp;
+    for (uint face = 0; face < lhs.get_num_faces(); face++)
+    {
+        for (uint level = 0; level < lhs.get_num_levels(); level++)
+        {
+            image_u8 *lhs_img = lhs.get_level_image(face, level, lhs_temp);
+            image_u8 *rhs_img = rhs.get_level_image(face, level, rhs_temp);
+            if (*lhs_img != *rhs_img)
+                return false;
+        }
+    }
+
+    return true;
+}
+} // namespace crnlib
+
+namespace btu::tex {
+using namespace crnlib;
+
+void CrunchTexture::set(mipmapped_texture &&tex) noexcept
+{
+    tex_ = std::move(tex);
+}
+
+auto CrunchTexture::get() noexcept -> mipmapped_texture &
+{
+    return tex_;
+}
+
+auto CrunchTexture::get() const noexcept -> const mipmapped_texture &
+{
+    return tex_;
+}
+
+auto CrunchTexture::get_dimension() const noexcept -> Dimension
+{
+    return {tex_.get_width(), tex_.get_height()};
+}
+
+auto CrunchTexture::get_load_path() const noexcept -> const Path &
+{
+    return load_path_;
+}
+
+void CrunchTexture::set_load_path(Path path) noexcept
+{
+    load_path_ = std::move(path);
+}
+
+auto CrunchTexture::get_texture_type() const noexcept -> const TextureType
+{
+    const auto filename = load_path_.filename().u8string();
+
+    const auto guessed = guess_texture_type(filename);
+    if (!guessed)
+        return TextureType::Diffuse;
+    else
+        return guessed.value();
+}
+
+auto CrunchTexture::get_format_as_dxgi() const noexcept -> const DXGI_FORMAT
+{
+    switch (tex_.get_format())
+    {
+        case PIXEL_FMT_DXT1:
+        case PIXEL_FMT_DXT1A: return DXGI_FORMAT_BC1_UNORM_SRGB;
+        case PIXEL_FMT_DXT2:
+        case PIXEL_FMT_DXT3: return DXGI_FORMAT_BC2_UNORM_SRGB;
+        case PIXEL_FMT_DXT4:
+        case PIXEL_FMT_DXT5: return DXGI_FORMAT_BC3_UNORM_SRGB;
+        case PIXEL_FMT_DXT5A: return DXGI_FORMAT_BC4_UNORM;
+        case PIXEL_FMT_3DC: return DXGI_FORMAT_BC5_UNORM;
+        case PIXEL_FMT_R8G8B8: return DXGI_FORMAT_B8G8R8X8_UNORM_SRGB;
+        case PIXEL_FMT_A8R8G8B8: return DXGI_FORMAT_B8G8R8A8_UNORM_SRGB;
+        default: return DXGI_FORMAT_UNKNOWN;
+    }
+}
+
+auto load_crunch(Path path) noexcept -> tl::expected<CrunchTexture, Error>
+{
+    CrunchTexture tex;
+    tex.set_load_path(std::move(path));
+
+    mipmapped_texture tex_;
+
+    const auto load_path = tex.get_load_path().string();
+
+    texture_file_types::format src_file_format = texture_file_types::determine_file_format(load_path.c_str());
+
+    const auto success = tex_.read_from_file(load_path.c_str(), src_file_format);
+    if (!success)
+    {
+        // Crunch saves loading errors in the texture object as a string.
+        // Also it is being cleared in a weird fashion so not super helpful.
+        return tl::make_unexpected(error_from_hresult(E_UNEXPECTED));
```

Comment:
Is there really no interesting information in the error? Even something vague is better than "unexpected"

Reply:
Tried to improve it, but looking at Crunch source, even if it saves the error information, the string is constantly being cleared in a weird fashion it is hard to figure out anything better than `ERROR_READ_FAULT`.